syntax = "proto3";

service Emulator {
    rpc RunNumInstructions (RunInstructionsRequest) returns (RunInstructionsReply) {}
}

message RunInstructionsRequest {
    int32 numInstructions = 1;
}

message MemoryAccess {
    uint64 vAddr = 1;
    uint32 size = 2;
    bool isStore = 3;
}

message RunInstructionsReply {
    repeated Instruction instructions = 1;

    message Instruction {
        bytes instruction = 1;
        bytes opCode = 2;
        optional MemoryAccess memoryAccess = 3;
    }
}
